openshift-201/rh201-pod-auto-scale.md

```diff
@@ -0,0 +1,283 @@
+
+Edit the load-test deployment requests environment variable which will re-trigger the deployment to start a load-test pod that will send traffic to the hello-world-nginx pod. You should see the number of pods increase as the CPU metrics grow.
+
+To get information about horizontal pod autoscaler resources in the current project, use the oc get command.
+
+```yaml
+oc get hpa 
+NAME                REFERENCE                      TARGETS    MINPODS   MAXPODS   REPLICAS
+hello-world-nginx   Deployment/hello-world-nginx   600%/80%   1         5          4        
+```
+
+The TARGETS column is the metrics on pods as a percentage of cpu request at current/target.
+
+The horizontal pod autoscaler initially has a value of `<unknown>` in the TARGETS column. It might take up to five minutes before <unknown> changes to display a percentage for current usage.
+
+A persistent value of `<unknown>` in the TARGETS column might indicate that the deployment does not define resource requests for the metric. The horizontal pod autoscaler will not scale these pods.
+
+### API Versions 
+
+There are different API versions for autoscaling v1 just works with CPU metrics. The v2beta2 API handles more options and metrics including cpu and memory.
+
+|Metric | Description | API version|
+|---|---|----|
+|CPU utilization| Number of CPU cores used. Can be used to calculate a percentage of the pod’s requested CPU.| autoscaling/v1, autoscaling/v2beta2|
+|Memory utilization| Amount of memory used. Can be used to calculate a percentage of the pod’s requested memory.| autoscaling/v2beta2|
+
+You can check the autoscaling API versions available in the cluster.
+
+```
+oc api-versions | grep autoscaling
+```
+
+The `oc autoscale` command will create a v1 type autoscaler. You can view with the hpa details with an `oc get hpa` command.
+
+```yaml
+oc get hpa -o yaml
+apiVersion: autoscaling/v1
+kind: HorizontalPodAutoscaler
+metadata:
+  name: hello-world-nginx
+spec:
+  maxReplicas: 5
+  minReplicas: 1
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: hello-world-nginx
+  targetCPUUtilizationPercentage: 80
+```
+
+To create a v2beta2 autoscaler you need to define in a yaml.
+
+```yaml
+apiVersion: autoscaling/v2beta2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: hello-world-nginx-mem-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: hello-world-nginx
+  minReplicas: 1
+  maxReplicas: 5
+  metrics:
+  - type: Resource
+    resource:
+      name: memory
+      target:
+        type: AverageValue
+        averageValue: 30Mi
+```
+
+Under metrics you can set type to AverageValue and specify averageValue memory.
+
+You can also specify Utilization in the metrics section of the v2beta2 HPA.
+
+```yaml
+metrics: 
+  - type: Deployment
```

Comment:
Should this be 'Resource' rather than Deployment?

Reply:
Yes! Good catch!